pkg/ovs/openflow/ofctrl_action.go

```diff
@@ -272,25 +293,38 @@ func (a *ofFlowAction) SetVLAN(vlanID uint16) FlowBuilder {
 
 // LoadARPOperation is an action to load data to NXM_OF_ARP_OP field.
 func (a *ofFlowAction) LoadARPOperation(value uint16) FlowBuilder {
-	loadAct, _ := ofctrl.NewNXLoadAction(NxmFieldARPOp, uint64(value), openflow13.NewNXRange(0, 15))
+	loadAct := &ofctrl.SetARPOpAction{Value: value}
 	a.builder.ApplyAction(loadAct)
 	return a.builder
 }
 
 // LoadRange is an action to load data to the target field at specified range.
 func (a *ofFlowAction) LoadRange(name string, value uint64, rng *Range) FlowBuilder {
-	loadAct, _ := ofctrl.NewNXLoadAction(name, value, rng.ToNXRange())
-	if a.builder.ofFlow.Table != nil && a.builder.ofFlow.Table.Switch != nil {
-		loadAct.ResetFieldLength(a.builder.ofFlow.Table.Switch)
+	hasMask := rng != nil
+	field, _ := openflow15.FindFieldHeaderByName(name, hasMask)
+	var valueBytes, maskBytes []byte
+	valueData := value
+	if hasMask {
+		mask := ^uint64(0) >> (64 - rng.Length()) << rng.Offset()
+		binary.BigEndian.PutUint64(maskBytes, mask)
+		field.Mask = util.NewBuffer(maskBytes)
+		valueData = valueData << rng.Offset()
 	}
-	a.builder.ApplyAction(loadAct)
-	return a.builder
+	binary.BigEndian.PutUint64(valueBytes, valueData)
+	field.Value = util.NewBuffer(valueBytes)
+	return a.setField(field)
 }
 
 func (a *ofFlowAction) LoadToRegField(field *RegField, value uint32) FlowBuilder {
-	name := field.GetNXFieldName()
-	loadAct, _ := ofctrl.NewNXLoadAction(name, uint64(value), field.rng.ToNXRange())
-	a.builder.ApplyAction(loadAct)
+	valueData := value
+	mask := uint32(0)
+	if field.rng != nil {
+		mask = ^mask >> (32 - field.rng.Length()) << field.rng.Offset()
+		valueData = valueData << field.rng.Offset()
```

Comment:
How about adding a function to generate mask and value? I saw more than one like this.

Reply:
I thought it was the basic bit shift operation. If a general function is wanted, the type of value should be `interface{}` (because it could be uint8, uint32, uint64), and we should check the type in the function to ensure the type case it correct, and we should also parse the bit size according to the value type. An alternative way is provide separate function for unit8, uint32, uint64, but I don't see the worth to do it.